### PR TITLE
[subscriptions] Refactor the subscription frequency selector code

### PR DESCRIPTION
**Pull request by smahbod2014**
# Description

This PR refactors the subscription frequency selector that we previously worked on for the plugin. Before, we copied the Product Page Checkout (PPC) code and made a new phtml file and wrote data to local storage. This refactor addresses the concerns brought up in [this doc](https://docs.google.com/document/d/1sWv2sMwTjFhJ1DM0xPbV1Dx2vcs5yVQYk5KJVs246kg/edit#), namely:

* We removed all the code in the `subscription_freq_selector_product_page.phtml` file except for the `div` that will be mounted (this can be further addressed if we want to remove the file entirely, but the div is positioned in a specific way at the moment, so I would rather do that separately)
* No more writing to local storage. Instead we leverage the `ajax:addToCart` event to call a function we've written in storm to persist subscription settings for the product.

#changelog Sean/m2 subs1

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 

- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my ticket link and provided a changelog message.

### CODE REVIEW

File: Block/Js.php [hunk 1]
```diff
@@ -91,6 +91,12 @@ public function __construct(
         $this->httpContext = $httpContext;
     }
 
+    public function getCurrentProduct() {
```
Comment (vitaliyreznikov) [inline Block/Js.php:94]: Could we add a check if we are on product page inside this function as well?

File: Block/Js.php [hunk 2]
```diff
@@ -91,6 +91,12 @@ public function __construct(
...
+        $objectManager = \Magento\Framework\App\ObjectManager::getInstance();
+        $product = $objectManager->get('Magento\Framework\Registry')->registry('current_product');
```
Comment (vitaliyreznikov) [inline Block/Js.php:96]: Using objectManager is not a proper magento way. @garmashdev could you please recommend how we can take the current project here?

File: view/frontend/web/js/bolt-api-driven-checkout.js [hunk 3]
```diff
@@ -1014,6 +1017,25 @@ define([
             );
         },
 
+        boltSetUpSubscriptionFrequencySelector: function() {
+            // deprecated, to be removed
```
Comment (vitaliyreznikov) [inline view/frontend/web/js/bolt-api-driven-checkout.js:1021]: Could you please add asana task with explanation here?

File: view/frontend/web/js/bolt-api-driven-checkout.js [hunk 4]
```diff
@@ -1077,9 +1102,39 @@ define([
...
+            return new Promise(resolve => {
+                setTimeout(() => { resolve('') }, millisec);
+            })
+        }
+
+        // When an item is added to an empty cart, the quoteMaskedId takes a moment to get populated, so wait for that
+        const maxAttempts = 3;
+        for (let i = 0; i < maxAttempts; i++) {
+            if (!BoltCheckoutApiDriven.quoteMaskedId) {
+                await sleep(1000);
```
Comment (vitaliyreznikov) [inline view/frontend/web/js/bolt-api-driven-checkout.js:1127]: Could we remove this sleep code?
I still think that moving wait logic to storm is the best approach, so we can have two API calls:
`window.BoltCheckout.persistSubscriptionFrequencySelection(quoteMaskedId) where quoteMaskedId may be null
`and
`window.BoltCheckout.setCartToken`

but as a simple middleground we can set local flag
```
if (BoltCheckoutApiDriven.quoteMaskedId) {
            window.BoltCheckout.persistSubscriptionFrequencySelection(BoltCheckoutApiDriven.quoteMaskedId);
} else {
        this.shouldCallPersistSubscriptionFrequencySelection = true
}
```
and after line
`BoltCheckoutApiDriven.quoteMaskedId = magentoCart.quoteMaskedId;`
add
```
if this.shouldCallPersistSubscriptionFrequencySelection {
 BoltCheckoutApiDriven.quoteMaskedId = magentoCart.quoteMaskedId;
 this.shouldCallPersistSubscriptionFrequencySelection = false
}
```

File: view/frontend/web/js/bolt-api-driven-checkout.js (hunk 3 above)
Comment (smahbod2014) [reply view/frontend/web/js/bolt-api-driven-checkout.js:1021]: I'm going to make a couple more changes to storm code, then I can remove this piece here.

Comment (vitaliyreznikov) [reply view/frontend/web/js/bolt-api-driven-checkout.js:1021]: Good! Let's add link to asana

**Comment by smahbod2014** [comment]
Thank you for the review. We've had a sudden change in priorities, so I'm going to have to pause work on this, but I promise I'll re-open this as soon as we're able to resume work.

**Pull request closed without merging**